File: darwin/flutter_compress_pro/Sources/flutter_compress_pro/FlutterCompressPlugin.swift
#if os(iOS)
import Flutter
#else
import FlutterMacOS
#endif
import Foundation

/// Bridges Flutter <-> the native pieces: `CompressionEngine` (transcode),
/// `MediaProbe`, `Thumbnailer`, `DownloadSaver`. Shared iOS / macOS (AVFoundation,
/// no FFmpeg).
public class FlutterCompressPlugin: NSObject, FlutterPlugin, FlutterStreamHandler {

  private let engine = CompressionEngine()
  private var eventSink: FlutterEventSink?
  private let workQueue = DispatchQueue(label: "flutter_compress_pro.work", qos: .userInitiated)

  public static func register(with registrar: FlutterPluginRegistrar) {
    #if os(iOS)
    let messenger = registrar.messenger()
    #else
    let messenger = registrar.messenger
    #endif
    let methodChannel = FlutterMethodChannel(
      name: "flutter_compress_pro/methods", binaryMessenger: messenger)
    let eventChannel = FlutterEventChannel(
      name: "flutter_compress_pro/progress", binaryMessenger: messenger)
    let instance = FlutterCompressPlugin()
    registrar.addMethodCallDelegate(instance, channel: methodChannel)
    eventChannel.setStreamHandler(instance)
    // publish() is what earns us `detachFromEngine(for:)`, so an engine teardown
    // can stop an in-flight export instead of leaving it running.
    registrar.publish(instance)
    instance.engine.onProgress = { [weak instance] payload in
      DispatchQueue.main.async { instance?.eventSink?(payload) }
    }
  }

  public func detachFromEngine(for registrar: FlutterPluginRegistrar) {
    engine.cancelAll()
    eventSink = nil
  }

  public func onListen(
    withArguments _: Any?, eventSink events: @escaping FlutterEventSink
  ) -> FlutterError? {
    eventSink = events
    return nil
  }

  public func onCancel(withArguments _: Any?) -> FlutterError? {
    eventSink = nil
    return nil
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    let args = call.arguments as? [String: Any] ?? [:]
    func str(_ key: String) throws -> String {
      guard let value = args[key] as? String else { throw Self.badArg(key) }
      return value
    }
    func map(_ key: String) throws -> [String: Any] {
      guard let value = args[key] as? [String: Any] else { throw Self.badArg(key) }
      return value
    }

    switch call.method {
    case "getVideoInfo":
      dispatch(result, ErrorCode.infoFailed) { try MediaProbe.videoInfo(path: try str("path")) }

    case "estimate":
      dispatch(result, ErrorCode.estimateFailed) {
        try self.engine.estimate(
          path: try str("path"), config: CompressionConfig(map: try map("config")))
      }

    case "compress":
      guard let config = try? CompressionConfig(map: map("config")),
        let id = try? str("id"), let path = try? str("path")
      else {
        result(
          FlutterError(
            code: ErrorCode.badArguments,
            message: "compress: missing id/path/config", details: nil))
        return
      }
      let outputDir = args["outputDir"] as? String, outputName = args["outputName"] as? String
      workQueue.async {
        self.engine.compress(
          id: id, path: path, config: config, outputDir: outputDir, outputName: outputName
        ) { outcome in
          DispatchQueue.main.async {
            switch outcome {
            case .success(let map): result(map)
            case .cancelled:
              result(
                FlutterError(
                  code: ErrorCode.cancelled, message: "Compression cancelled", details: nil))
            case .failure(let message):
              result(FlutterError(code: ErrorCode.compressFailed, message: message, details: nil))
            }
          }
        }
      }

    case "cancel":
      engine.cancel(id: args["id"] as? String)
      result(nil)

    case "isCompressing":
      result(engine.isCompressing())

    case "getThumbnail":
      dispatch(result, ErrorCode.thumbnailFailed) {
        try Thumbnailer.generate(
          dir: PluginFiles.cacheDir(), path: try str("path"),
          positionMs: (args["positionMs"] as? NSNumber)?.int64Value ?? 0,
          quality: (args["quality"] as? NSNumber)?.intValue ?? 80,
          maxWidth: (args["maxWidth"] as? NSNumber)?.intValue)
      }

    case "clearCache":
      PluginFiles.clearCache()
      result(nil)

    case "saveToDownloads":
      dispatch(result, ErrorCode.saveFailed) {
        try DownloadSaver.save(path: try str("path"), fileName: args["fileName"] as? String)
      }

    case "getImageInfo":
      dispatch(result, ErrorCode.imageInfoFailed) { try ImageEngine.info(path: try str("path")) }

    case "compressImageBytes":
      dispatch(result, ErrorCode.imageCompressFailed) {
        guard let data = args["bytes"] as? FlutterStandardTypedData else {
          throw Self.badArg("bytes")
        }
        return try ImageEngine.compressBytes(
          data: data.data, config: ImageConfig(map: try map("config")))
      }

    case "compressImage":
      dispatch(result, ErrorCode.imageCompressFailed) {
        try ImageEngine.compress(
          path: try str("path"),
          config: ImageConfig(map: try map("config")),
          outputDir: args["outputDir"] as? String,
          outputName: args["outputName"] as? String)
      }

    default:
      result(FlutterMethodNotImplemented)
    }
  }

  private func dispatch(
    _ result: @escaping FlutterResult, _ errorCode: String, _ block: @escaping () throws -> Any?
  ) {
    workQueue.async {
      let outcome: Result<Any?, Error>
      do { outcome = .success(try block()) } catch { outcome = .failure(error) }
      DispatchQueue.main.async {
        switch outcome {
        case .success(let value):
          result(value)
        case .failure(let error):
          result(FlutterError(code: errorCode, message: error.localizedDescription, details: nil))
        }
      }
    }
  }

  private static func badArg(_ key: String) -> NSError {
    NSError(
      domain: "flutter_compress_pro", code: 1,
      userInfo: [NSLocalizedDescriptionKey: "Missing or invalid argument: \(key)"])
  }
}

File: android/src/main/kotlin/com/compress/all/flutter_compress_pro/FlutterCompressPlugin.kt
package com.compress.all.flutter_compress_pro

import android.content.Context
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import kotlin.coroutines.CoroutineContext

/**
 * Bridges Flutter <-> the native pieces: [CompressionEngine] (transcode),
 * [MediaProbe], [Thumbnailer], [DownloadSaver].
 *
 * Only video transcoding runs on the main looper — Media3's `Transformer` must be
 * created and driven from a thread that has a [android.os.Looper]. Everything
 * else is dispatched to a background dispatcher and replies on the main thread
 * (see [dispatch]).
 */
class FlutterCompressPlugin :
    FlutterPlugin,
    MethodCallHandler,
    EventChannel.StreamHandler {

    private lateinit var methodChannel: MethodChannel
    private lateinit var eventChannel: EventChannel
    private lateinit var context: Context

    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Main.immediate)
    private var engine: CompressionEngine? = null
    private var imageEngine: ImageEngine? = null
    private var eventSink: EventChannel.EventSink? = null

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        context = binding.applicationContext
        methodChannel = MethodChannel(binding.binaryMessenger, "flutter_compress_pro/methods")
        methodChannel.setMethodCallHandler(this)
        eventChannel = EventChannel(binding.binaryMessenger, "flutter_compress_pro/progress")
        eventChannel.setStreamHandler(this)
        engine = CompressionEngine(context) { eventSink?.success(it) }
        imageEngine = ImageEngine(context)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        methodChannel.setMethodCallHandler(null)
        eventChannel.setStreamHandler(null)
        engine?.cancelAll()
        scope.cancel()
    }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        eventSink = events
    }

    override fun onCancel(arguments: Any?) {
        eventSink = null
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        val eng = engine ?: return result.error(ErrorCode.NO_ENGINE, "Engine not initialized", null)
        when (call.method) {
            "getVideoInfo" ->
                dispatch(result, ErrorCode.INFO_FAILED) { MediaProbe.videoInfo(call.str("path")) }

            "estimate" ->
                dispatch(result, ErrorCode.ESTIMATE_FAILED) { eng.estimate(call.str("path"), call.config()) }

            // Media3's Transformer must be created and driven from a Looper thread.
            "compress" -> dispatch(result, ErrorCode.COMPRESS_FAILED, Dispatchers.Main.immediate) {
                eng.compress(
                    call.str("id"), call.str("path"), call.config(),
                    call.argument<String>("outputDir"),
                    call.argument<String>("outputName"),
                )
            }

            "cancel" -> {
                eng.cancel(call.argument<String>("id"))
                result.success(null)
            }

            "isCompressing" -> result.success(eng.isCompressing())

            "getThumbnail" -> dispatch(result, ErrorCode.THUMBNAIL_FAILED) {
                Thumbnailer.generate(
                    context.compressCacheDir(), call.str("path"),
                    call.long("positionMs", 0), call.int("quality", 80),
                    call.argument<Number>("maxWidth")?.toInt(),
                )
            }

            // Deleting a cache full of multi-hundred-MB transcodes is real IO.
            "clearCache" -> dispatch(result, ErrorCode.SAVE_FAILED, Dispatchers.IO) {
                context.clearCompressCache()
                null
            }

            "saveToDownloads" -> dispatch(result, ErrorCode.SAVE_FAILED, Dispatchers.IO) {
                DownloadSaver.save(context, call.str("path"), call.argument<String>("fileName"))
            }

            // ---- images ----
            "getImageInfo" -> dispatch(result, ErrorCode.IMAGE_INFO_FAILED) {
                requireNotNull(imageEngine).info(call.str("path"))
            }

            "compressImageBytes" -> dispatch(result, ErrorCode.IMAGE_COMPRESS_FAILED) {
                val bytes = call.argument<ByteArray>("bytes")
                    ?: throw BadArgumentException("bytes")
                requireNotNull(imageEngine).compressBytes(bytes, call.imageConfig())
            }

            "compressImage" -> dispatch(result, ErrorCode.IMAGE_COMPRESS_FAILED) {
                requireNotNull(imageEngine).compress(
                    call.str("path"),
                    call.imageConfig(),
                    call.argument<String>("outputDir"),
                    call.argument<String>("outputName"),
                )
            }

            else -> result.notImplemented()
        }
    }

    /**
     * Run [block] on [ctx], then deliver its value (or a typed error) on the
     * main thread, as `MethodChannel.Result` requires.
     *
     * Defaults to a background dispatcher: only Media3's `Transformer` needs the
     * main looper, and everything else here (image encode, probes, thumbnails,
     * MediaStore copies) is blocking CPU/IO work that would otherwise ANR.
     */
    private fun dispatch(
        result: Result,
        errorCode: String,
        ctx: CoroutineContext = Dispatchers.Default,
        block: suspend () -> Any?,
    ) {
        scope.launch {
            runCatching { withContext(ctx) { block() } }
                .onSuccess { result.success(it) }
                .onFailure { e ->
                    when (e) {
                        is CompressionCancelledException ->
                            result.error(ErrorCode.CANCELLED, e.message, null)
                        is BadArgumentException ->
                            result.error(ErrorCode.BAD_ARGUMENTS, e.message, null)
                        is PermissionDeniedException ->
                            result.error(ErrorCode.PERMISSION_DENIED, e.message, null)
                        // Keep the native detail: it's the only clue to the origin.
                        else -> result.error(errorCode, e.message, e.stackTraceToString())
                    }
                }
        }
    }

    // ---- argument accessors ------------------------------------------------

    /** Missing/mistyped args must surface as [ErrorCode.BAD_ARGUMENTS], matching
     *  iOS — `!!` here would leak an NPE as whatever the caller's error code is. */
    private fun MethodCall.str(key: String): String =
        argument<String>(key) ?: throw BadArgumentException(key)
    private fun MethodCall.int(key: String, default: Int): Int =
        (argument<Number>(key) ?: default).toInt()
    private fun MethodCall.long(key: String, default: Long): Long =
        (argument<Number>(key) ?: default).toLong()

    private fun MethodCall.config(): CompressionConfig =
        CompressionConfig.fromMap(
            argument<Map<String, Any?>>("config") ?: throw BadArgumentException("config"),
        )

    private fun MethodCall.imageConfig(): ImageConfig =
        ImageConfig.fromMap(
            argument<Map<String, Any?>>("config") ?: throw BadArgumentException("config"),
        )
}
